### PR TITLE
[IMP] web: field service

### PR DESCRIPTION
**Pull request by Polymorphe57**

### CODE REVIEW

File: addons/web/static/src/core/utils/cache.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ export class Cache {
         this.getKey = getKey;
         this.getValue = getValue;
     }
-    read(...path) {
+    _getKey(...path) {
```
Comment (Polymorphe57) [inline addons/web/static/src/core/utils/cache.js:9]: _getCacheKey?

File: addons/web/static/src/core/utils/cache.js [hunk 2]
```diff
@@ -17,6 +17,17 @@ export class Cache {
             }
             key = path[path.length - 1];
         }
+        return { cache, key };
+    }
+    clear(...path) {
+        const { cache, key } = this._getCacheAndKey(...path);
+        delete cache[key];
+    }
+    clearAll() {
```
Comment (ged-odoo) [inline addons/web/static/src/core/utils/cache.js:26]: rename to invalidate

File: addons/web/static/src/core/field_service.js [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        async function _loadPath(resModel, fieldDefs, names) {
+            const [name, ...remaningNames] = names;
+            const modelsInfo = [{ resModel, fieldDefs }];
+
+            const fieldDef = fieldDefs[name];
+            if (name !== "*" && !fieldDef) {
+                return { isInvalid: "path", names, modelsInfo };
+            }
+
+            if (!remaningNames.length) {
```
Comment (ged-odoo) [inline addons/web/static/src/core/field_service.js:68]: remainingNames

File: addons/web/static/tests/core/field_service_tests.js [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+
+    const env = await makeTestEnv({ serverData, mockRPC });
+
+    await env.services.field.loadFields("tortoise");
+    await env.services.field.loadFields("tortoise");
+
+    assert.verifySteps(["fields_get"]);
+});
+
+QUnit.test("store loadFields calls in cache when failed", async (assert) => {
```
Comment (aab-odoo) [inline addons/web/static/tests/core/field_service_tests.js:185]: does not store

File: addons/web/static/tests/core/field_service_tests.js [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+                        tortoise_ids: { type: "one2many", string: "Turtles", relation: "tortoise" },
+                    },
+                },
+            },
+        };
+        serviceRegistry.add("field", fieldService);
+    },
+});
+
+QUnit.test("loadPath", async (assert) => {
```
Comment (aab-odoo) [inline addons/web/static/tests/core/field_service_tests.js:46]: in general we define tests inside the definition of the module, not after (not sure how it behaves, maybe it works fine)

File: addons/web/static/tests/views/fields/domain_field_tests.js [hunk 6]
```diff
@@ -815,7 +818,7 @@ QUnit.module("Fields", (hooks) => {
         assert.verifySteps(["/web/webclient/load_menus"]);
 
         await doAction(webClient, 1);
-        assert.verifySteps(["/web/action/load", "get_views", "read", "search_count"]);
+        assert.verifySteps(["/web/action/load", "get_views", "read", "search_count", "fields_get"]);
```
Comment (aab-odoo) [inline addons/web/static/tests/views/fields/domain_field_tests.js:821]: extra rpc?

Comment (Polymorphe57) [reply addons/web/static/tests/views/fields/domain_field_tests.js:821]: before, the view service did put the fields it gets from get_views in cache. Here we have to separated caches managed by 2 two different services. Gery told me that it's fine.

File: addons/web/static/tests/core/field_service_tests.js (hunk 5 above)
Comment (Polymorphe57) [reply addons/web/static/tests/core/field_service_tests.js:46]: It works fine

**Review by aab-odoo — APPROVED** [review]

**Comment by aab-odoo** [comment]
robodoo r+

**Comment by aab-odoo** [comment]
robodoo rebase-ff r+

**Pull request closed without merging**